### PR TITLE
Cache item types for inventory; huge crafting GUI speedup

### LINKED ISSUE #18811: Crafting inventory (structure, not UI) sorting by ID
Crafting inventory currently does a linear check for every single item it wants.
This is useful for abstract definitions of things, but also happens to be of linear complexity. Per every item - populating a craft menu requires checking every single ingredient for every single recipe on the page.
For example, if you have 100 non-stacking items and populate a page of 50 recipes with average of 2 ingredients, you need to do 1000 item checks. And that isn't even a big example - it's very easy to get a better one in the game.

A relatively simple way to fix that would be to bin inventory contents by item ID. This could be done on demand and stored in a `mutable` cache.
That is, a `std::map<item_id, std::vector<const item *>>` populated simply by going over all items and adding them to the list that has their type IDs as the key.

There is one problem here that needs solving first: the inability to have mixed AND/OR in recipes.
This functionality is actually used, using the subtype system for crafting tools.
Alternatively, a hack could be added: each item would "register" self both in its actual ID list and in the list for the subtype. This would slow things down by requiring ID checks, but would simplify writing and still easily offer speedups of up to 3 orders of magnitude.

Another problem is that currently the crafting system has some bugs. That's why this should probably wait a bit.

### PR DESCRIPTION
**Pull request by Coolthulhu**
Closes #18811

Master version works like this:
- For every recipe to display
- For every component in said recipe
- Search the cached inventory going linearly, checking for matches, until enough matches are found or entire inventory is processed

Version here:
- Go through the inventory once
- Note which item has what type, build a map of item types to lists of pointers to items of said type
- When checking inventory for items, only check the items from the list for the relevant type

While this smaller list can still be giant, like when you for some reason haul 10k rags into one spot, those rags will not be checked for until someone wants rags. And then only enough rags to meet the demand will be checked.

As a result, crafting bases which would kill the current master version due to overload of items of different types have only a tiny effect on performance. This effect is mostly due to the fact that many recipes are checked and many of them have a lot of components.

[CR] because:
- The way I achieved the "binning request" is pretty fragile - the result has to be cached somewhere accessible to `inventory::charges_of/amount_of`, heavily suggesting `inventory`. So this is cached in `inventory`. But then, `inventory` isn't perfectly encapsulated, meaning that the binning feature should not be used with any inventory that can change.
- I had to get rid of tool subtype feature/hack usage. I achieved it by unwrapping any subtyped tool used in any `requirements`. This makes crafting lists longer. I could also have missed a use case of subtype, though that's unlikely because I addressed all relevant lines given by `grep "subtype" -r src`.

### CODE REVIEW

**Comment by mutability** [comment]
It seems wrong for visitable<inventory>::amount_of / visitable<inventory>::charges_of to behave differently to the non-specialized versions. (get_binned_items says "NOTE: Assumes items will be used for crafting and may skip those which can't be!")

**Comment by mutability** [comment]
At the points where you set binned = false, also clear binned_items? (and probably should be factored out to an invalidation method)

**Comment by Coolthulhu** [comment]
> It seems wrong for visitable<inventory>::amount_of / visitable<inventory>::charges_of to behave differently to the non-specialized versions.

Makes sense, though I'm not sure if I can think of a counterexample here.
I'll probably need to change the crafting system to use some specialized charge counter then.

> At the points where you set binned = false, also clear binned_items? (and probably should be factored out to an invalidation method)

`binned_items` is always accessed through a getter which clears it anyway. There shouldn't be a case where someone uses `binned_items` without getting it through said getter.

**Comment by Coolthulhu** [comment]
I realized that this limitation is actually a non-issue: all the skipped items are those which would be skipped by `visit_items` anyway.

Should be ready now.

File: src/requirements.h [hunk 1]
```diff
@@ -237,6 +240,8 @@ struct requirement_data {
         static void check_consistency( const std::vector< std::vector<T> > &vec,
                                        const std::string &display_name );
         template<typename T>
+        static void finalize( std::vector< std::vector<T> > &vec );
```
Comment (mugling) [inline src/requirements.h]: Why is this a template when you only need one variant of it?

File: src/item_factory.h [hunk 2]
```diff
@@ -313,6 +315,12 @@ class Item_factory
         void add_actor( iuse_actor *ptr );
 
         std::map<itype_id, migration> migrations;
+
+        /**
+         * Contains the tool subtype mappings for crafing (ie. mess kit is a hotplate etc.).
+         * This is should be obsoleted when @ref requirement_data allows AND/OR nesting.
```
Comment (mugling) [inline src/item_factory.h]: Or when we complete the `using` requirement sets?

**Comment by mugling** [comment]
This looks reasonable but review of the binning code might be be tricky. Can you factor out a revision of the subtype part as that's an otherwise elegant fix for #17517

File: src/item_factory.h (hunk 2 above)
Comment (Coolthulhu) [inline src/item_factory.h]: I'd say those are equivalent.

**Comment by Coolthulhu** [comment]
I could try, though it could cause a rather significant regression in performance without binning.

**Comment by mugling** [comment]
> I could try, though it could cause a rather significant regression in performance without binning.

Correct is more important than performant. I'd prefer to test binning separately as we need to be careful it doesn't introduce subtle bugs as I don't think its completely encapsulated?

**Comment by mugling** [comment]
Dependent upon #19052 which currently has load errors

**Pull request merged**